src/System.IO.Pipelines.Networking.Tls/OpenSslSecurityContext.cs

```diff
@@ -0,0 +1,141 @@
+    {
+        //Current fixed block size (4k - 1 64 byte cacheline, should be from the pipeline factory in the future
+        internal const int BlockSize = 1024 * 4 - 64; 
+
+        private readonly string _hostName;
+        private readonly PipelineFactory _pipelineFactory;
+        private readonly bool _isServer;
+        private InteropKeys.PK12Certifcate _certifcateInformation;
+        private byte[] _alpnSupportedProtocolsBuffer;
+        private GCHandle _alpnHandle;
+        internal IntPtr SslContext;
+        internal Interop.alpn_cb AlpnCallback;
+        private ApplicationProtocols.ProtocolIds _alpnSupportedProtocols;
+
+        public OpenSslSecurityContext(PipelineFactory pipelineFactory, string hostName, bool isServer, string pathToPfxFile, string password)
+            : this(pipelineFactory, hostName, isServer, pathToPfxFile, password, 0)
+        {
+        }
+
+        public OpenSslSecurityContext(PipelineFactory pipelineFactory, string hostName, bool isServer, string pathToPfxFile, string password, ApplicationProtocols.ProtocolIds alpnSupportedProtocols)
+        {
+            if (isServer && string.IsNullOrEmpty(pathToPfxFile))
+            {
+                throw new ArgumentException("We need a certificate to load if you want to run in server mode");
+            }
+
+            InteropCrypto.Init();
+            _hostName = hostName;
+            _pipelineFactory = pipelineFactory;
+            _isServer = isServer;
+            _alpnSupportedProtocols = alpnSupportedProtocols;
+
+            if (!string.IsNullOrWhiteSpace(pathToPfxFile))
+            {
+                InteropBio.BioHandle fileBio = new InteropBio.BioHandle();
+                try
+                {
+                    fileBio = InteropBio.BIO_new_file_read(pathToPfxFile);
+                    //Now we pull out the private key, certificate and Authority if they are all there
+                    _certifcateInformation = new InteropKeys.PK12Certifcate(fileBio, password);
+                }
+                finally
+                {
+                    fileBio.FreeBio();
+                }
+            }
+            SetupContext();
+            SetupAlpn();
+        }
+
+        public bool IsServer => _isServer;
+        internal InteropKeys.PK12Certifcate CertificateInformation => _certifcateInformation;
+        internal IntPtr AplnBuffer => _alpnHandle.AddrOfPinnedObject();
+        internal int AplnBufferLength => _alpnSupportedProtocolsBuffer?.Length ?? 0;
+
+        private unsafe void SetupAlpn()
+        {
+            if (_alpnSupportedProtocols > 0)
+            {
+                //We need to get a buffer for the ALPN negotiation and pin it for sending to the lower API
+                _alpnSupportedProtocolsBuffer = ApplicationProtocols.GetBufferForProtocolId(_alpnSupportedProtocols, false);
+                _alpnHandle = GCHandle.Alloc(_alpnSupportedProtocolsBuffer, GCHandleType.Pinned);
```

Comment:
Why pin on demand.. normal use case for Tls is, App starts up, you make a security context(the point this is pinned) and then loads of connections come and go, if you are pinning/unpinning during server load and many connections you are more likely to fragment memory. Could consider malloc outside of the GC.